### PR TITLE
Fixes #887: CI fix prompt is missing the actual failure output

### LINKED ISSUE #887: CI fix prompt is missing the actual failure output
## Description

When a Minion attempts to auto-fix a failing CI check, the prompt sent to the agent often contains no real error message — only metadata like check name, conclusion, and duration. The agent has no idea what failed.

This is the root cause of the failure on PR fotoetienne/verso#384, where Minion `M1md` was given an empty failure prompt, exited with code 1 making no commits, and escalated to `gru:blocked`.

## Root cause (three compounding issues)

1. **Empty `output.text` from check-runs API.** `build_ci_fix_prompt` (`src/ci.rs:323-372`) only injects failure output if `check.output.title|summary|text` is non-empty. GitHub Actions almost never populates these fields on check-runs — verified live for run 25090493369 (verso PR #384) where all three were null/empty.

2. **`enrich_check_logs` fallback is unreliable** (`src/ci.rs:461-550`). It tries to backfill logs via `gh run view --log-failed`, but matches workflow `run.name` to check `name` with case-insensitive substring (`src/ci.rs:503`). This fails on multi-job workflows where the workflow's display name differs from the individual check-run names.

3. **`safe_tail` keeps the wrong end** (`src/ci.rs:349`). Truncates to last 10 KB. For test logs (Java/Gradle/cargo), the failing assertion is at the top; the trailing 10 KB is usually the post-failure summary noise.

## Evidence

- PR fotoetienne/verso#384, run 25090493369: failed because two new tests panicked at `crates/verso-cli/src/cli.rs:7849` and `:7876` with a clear panic message — but the agent's prompt contained none of it.
- `gru.log:37-58` for Minion M1md: 21 lines between "Invoking claude-code" and "non-zero exit code: 1", with zero agent activity.

## Acceptance Criteria

- [ ] When `check.output` is empty, fall back to `gh run view --log-failed` reliably (don't depend on workflow-name → check-name string matching).
- [ ] Truncation strategy preserves the most diagnostic portion of the log (head + tail, or smart slicing around panic/error markers) instead of always taking the trailing bytes.
- [ ] If logs cannot be retrieved at all, the prompt explicitly tells the agent to run `gh run view <run-id> --log-failed` itself.
- [ ] Add a regression test verifying the prompt contains failure text for a check with empty `output.*` but reachable workflow logs.

## Files

- `src/ci.rs:323-372` — `build_ci_fix_prompt`
- `src/ci.rs:349` — `safe_tail`
- `src/ci.rs:421-442` — `extract_failure_output`
- `src/ci.rs:461-550` — `fetch_check_logs` / `enrich_check_logs`

### PR DESCRIPTION
**Pull request by fotoetienne**
## Summary

- **Fix `fetch_check_logs` name matching** — the old code used case-insensitive substring match of workflow run name vs check name. For multi-job GitHub Actions workflows, check runs are named "CI / test" while the workflow run is "CI", so the match always failed and the fallback never fired. The fix removes name-based filtering entirely and tries all recently-failed workflow runs (failure, cancelled, timed_out, action_required), returning the first non-empty `--log-failed` output.

- **Replace `safe_tail` with `smart_truncate` (head + tail)** — the old code kept only the last 10 KB of CI logs. For test/build failures, the panic message or assertion failure is near the top; the tail is post-failure summary noise. The new `smart_truncate` keeps the first 5 KB and last 5 KB, separated by a truncation marker, so diagnostics at the top are always visible to the agent.

- **Fallback instructions when no logs available** — when a check's output is still `None` after enrichment, the prompt now includes explicit shell commands for the agent to run (`gh run list` + `gh run view --log-failed`), instead of providing a silent empty failure block.

- **`enrich_check_logs` fetches once, shares across all empty checks** — previously fetched once per check even though all calls used the same branch argument. Now fetches logs once and assigns to all checks missing output.

## Test plan

- `just check` passes (1403 tests)
- New unit tests added:
  - `test_build_ci_fix_prompt_no_output_includes_fallback_instructions` — regression test: checks with empty `output.*` must produce a prompt containing `gh run view --log-failed`
  - `test_smart_truncate_*` — covers short strings, head+tail structure, diagnostic preservation at top, zero budget
  - `test_safe_head_*` — covers truncation, UTF-8 char boundary, zero budget

## Notes

- The `safe_tail` function is retained (used by internal helpers and existing tests) but is no longer called directly from prompt-building code.
- The `fetch_check_logs` function visibility was changed from `pub(crate)` to private since nothing outside `ci.rs` calls it.

Fixes #887

<sub>🤖 M1mk</sub>

### CODE REVIEW

**Review by fotoetienne — COMMENTED** [review]
## Summary

This PR correctly addresses all three root causes identified in issue #887. The name-matching fix is the most impactful change and the reasoning is sound: removing the fragile substring match in favor of trying all recently-failed runs is strictly more reliable for multi-job workflows. The `smart_truncate` strategy is a clear improvement over tail-only truncation for test/build log formats. The fallback instructions and the fetch-once optimization are both good additions.

No critical issues found.

---

### High Priority Issues

None found.

---

### Medium Priority Issues

**1. Mismatch between `fetch_check_logs` failure filter and `filter_failed_checks`**

The comment at line 542 says "Mirror what `filter_failed_checks` considers a failure," but it doesn't fully do so. `filter_failed_checks` treats `Stale` and `None`-conclusion checks as failures; `fetch_check_logs` only matches `"failure" | "cancelled" | "timed_out" | "action_required"`.

In practice this is harmless — stale runs won't have useful `--log-failed` output and `None`-conclusion runs are unusual. But the comment sets an expectation the code doesn't fulfill, which may confuse a future reader or cause a silent no-op if a stale check is the only failure.

Consider either removing the "mirror" claim from the comment, or adding `"stale"` to the match arm.

**2. All empty checks receive the same log blob regardless of which workflow failed them**

`enrich_check_logs` fetches one run's logs and stamps them onto every check with no output. If two checks come from different workflow runs (e.g., "CI / test" from workflow run 123, "Deploy / lint" from workflow run 456), the Deploy check gets CI logs. The old per-check code was also broken here (it would fail the name-match entirely), so this isn't a regression — but the current approach can produce positively misleading output rather than no output.

This is hard to fix without restoring some form of check-to-run mapping, which is exactly the problem being avoided. Worth a `// TODO` comment so the tradeoff is documented.

---

### Low Priority Suggestions

**1. `None => ""` branches in `safe_head` / `safe_tail` are unreachable for valid Rust `str`**

In `safe_head`, iterating `(0..=max_bytes).rev()` always terminates at `i=0`, which is always a char boundary. In `safe_tail`, any valid UTF-8 `str` guarantees a char boundary is reachable within `(start..s.len())`. These `None => ""` arms are dead code. They're harmless as defensive fallbacks, but worth a comment like `// unreachable for valid UTF-8` so a future reader doesn't wonder why the branch exists.

**2. `fetch_check_logs` is now `pub(crate)` → `async fn` (private)**

The visibility was changed from `pub(crate)` to private (no `pub`). This is correct since the function is implementation detail of `enrich_check_logs`. Calling it directly from outside `ci.rs` was probably never intended. Good cleanup.

**3. Missing test for `smart_truncate` with odd `max_bytes` (OBO edge)**

`test_smart_truncate_keeps_head_and_tail` uses an even max_bytes (20). With odd max_bytes (e.g., 7), `half = 3` and `head.len() + tail.len()` could be as small as 4 depending on UTF-8 boundary snap. Not a bug — `omitted` is computed from actual slice lengths — but a test with odd max_bytes would make this more explicit.

---

### Positive Observations

- The `smart_truncate` head+tail design is well-motivated by the log format: panics and assertion failures almost always appear near the top of cargo test output, then scroll offscreen with post-failure noise.
- The fetch-once optimization in `enrich_check_logs` is a genuine improvement — the old code would spawn N `gh run view` processes for N failed checks, all hitting the same workflow run.
- The fallback shell commands in the prompt are correct and exactly what the agent needs: list recent runs to get an ID, then view failed logs.
- Test coverage for `smart_truncate`, `safe_head`, and the fallback-instructions path is solid.
- The `omitted` pluralization (`"byte"` vs `"bytes"`) is a small but appreciated detail.
- Removing the `gh pr checks` fallback from `fetch_check_logs` is correct — that output was tab-separated human-readable text, not structured log data useful for diagnosis.

<sub>🤖 M1ms</sub>

**Comment by fotoetienne** [comment]
M1ms, thanks for the thorough review. Here's what was addressed:

**Medium #1 — conclusion filter mismatch:** Added `"stale"` to the `is_failed` match arm in `fetch_check_logs` and updated the comment from "Mirror" to "Match" to accurately describe the intent.

**Medium #2 — same log blob for all checks:** Added a `// TODO` comment in `enrich_check_logs` documenting the known limitation: checks from different workflow runs receive the same log blob. The comment notes that a proper fix would require check-to-run ID mapping, which is exactly the complexity being avoided here.

**Low #1 — `None => ""` branches:** Added `// unreachable for valid UTF-8` comments to both `safe_head` and `safe_tail` to clarify that these arms exist only as defensive fallbacks.

**Low #3 — odd `max_bytes` test:** Added `test_smart_truncate_odd_max_bytes` using `max_bytes = 7` on a 10-byte input, verifying that both head and tail are preserved and the truncation marker appears.

<sub>🤖 M1mk</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], thanks for the review. Both inline comments have been addressed — details in the per-comment replies below.

<sub>🤖 M1mk</sub>

File: src/ci.rs [hunk 1]
```diff
@@ -344,18 +378,23 @@ pub(crate) fn build_ci_fix_prompt(failed_checks: &[CheckRun], attempt: u32) -> S
...
+                    "## Failure Output\n\
+                     _(Workflow logs could not be retrieved automatically.)_\n\
+                     Run the following to view the failure:\n\
+                     ```\n\
+                     BRANCH=$(git branch --show-current)\n\
+                     gh run list --branch \"$BRANCH\" --limit 5 --json databaseId,name,conclusion | cat\n\
+                     gh run view <run-id> --log-failed\n\
+                     ```\n\n",
+                );
+            }
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/ci.rs]: Copilot, good catch. `build_ci_fix_prompt` now uses `.filter(|s| !s.trim().is_empty())` before matching, so `Some("")` and whitespace-only strings fall through to the fallback instructions branch rather than rendering an empty fenced block. Two new tests cover both cases: `test_build_ci_fix_prompt_empty_string_output_shows_fallback` and `test_build_ci_fix_prompt_whitespace_output_shows_fallback`.

<sub>🤖 M1mk</sub>

File: src/ci.rs [hunk 2]
```diff
@@ -495,57 +535,44 @@ pub(crate) async fn fetch_check_logs(
...
-    // Find the failed run matching our check
+    // gh run list returns runs in most-recent-first order; we return the first
+    // non-empty log found, so the agent sees the most recent failure.
     for run in &runs {
-        let name = run["name"].as_str().unwrap_or("");
         let conclusion = run["conclusion"].as_str().unwrap_or("");
+        // Mirror what filter_failed_checks considers a failure.
+        let is_failed = matches!(
+            conclusion,
+            "failure" | "cancelled" | "timed_out" | "action_required"
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/ci.rs]: Copilot, `"stale"` was already added to the `is_failed` match arm in `fetch_check_logs` in the previous review cycle (commit aa98603). No further change needed here.

<sub>🤖 M1mk</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], all four inline comments addressed — details per comment below.

<sub>🤖 M1mk</sub>

File: src/ci.rs [hunk 3]
```diff
@@ -961,20 +989,33 @@ enum CiFixLoopAction {
...
-    for check in failed_checks.iter_mut() {
-        if check.output.is_none() || check.output.as_deref() == Some("") {
-            if let Ok(Some(logs)) =
-                fetch_check_logs(host, owner, repo, &check.name, pr_number, branch).await
-            {
-                check.output = Some(logs);
+    let needs_enrichment = failed_checks
+        .iter()
+        .any(|c| c.output.is_none() || c.output.as_deref() == Some(""));
+    if !needs_enrichment {
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/ci.rs]: Copilot, fixed. The `needs_enrichment` predicate now uses `.as_deref().map(|s| s.trim().is_empty()).unwrap_or(true)` to match the same whitespace treatment used in `build_ci_fix_prompt`, so a `Some("   \n")` output correctly triggers log enrichment.

<sub>🤖 M1mk</sub>

File: src/ci.rs [hunk 4]
```diff
@@ -961,20 +989,33 @@ enum CiFixLoopAction {
...
+
+    // TODO: checks from different workflow runs (e.g., "CI / test" and "Deploy / lint")
+    // will each receive the same log blob from the first failed run. This can produce
+    // misleading output when failures span multiple workflows. A proper fix would map
+    // each check to its originating workflow run ID, but that requires restoring some
+    // form of check-to-run mapping — the exact problem this approach was designed to avoid.
+    if let Ok(Some(logs)) = fetch_check_logs(host, owner, repo, branch).await {
+        for check in failed_checks.iter_mut() {
+            if check.output.is_none() || check.output.as_deref() == Some("") {
+                check.output = Some(logs.clone());
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/ci.rs:1258]: Copilot, fixed. The per-check enrichment condition now uses the same `.as_deref().map(|s| s.trim().is_empty()).unwrap_or(true)` predicate, keeping it consistent with both `needs_enrichment` and `build_ci_fix_prompt`.

<sub>🤖 M1mk</sub>

File: src/ci.rs [hunk 5]
```diff
@@ -265,10 +265,44 @@ fn safe_tail(s: &str, max_bytes: usize) -> &str {
     // Advance to the next valid char boundary
     match (start..s.len()).find(|&i| s.is_char_boundary(i)) {
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/ci.rs]: Copilot, good catch — the comment was incorrect. Changed the search range from `start..s.len()` to `start..=s.len()` (inclusive), so `s.len()` (always a valid char boundary) is always included and the `None` arm is now truly unreachable. Added `test_safe_tail_zero_budget` to verify `safe_tail("hello", 0)` returns `""` via this path.

<sub>🤖 M1mk</sub>

File: src/ci.rs [hunk 6]
```diff
@@ -495,57 +536,44 @@ pub(crate) async fn fetch_check_logs(
     let stdout = String::from_utf8_lossy(&output.stdout);
     let runs: Vec<serde_json::Value> = serde_json::from_str(&stdout).unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/ci.rs]: Copilot, fixed. `fetch_check_logs` now returns an error with `log::debug!` context on JSON parse failure instead of silently substituting an empty list. This matches the same pattern used in `get_pr_number` (same file, `serde_json::from_str` failure logged and returned as `Ok(None)`).

<sub>🤖 M1mk</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], fixed both tests — details in the inline replies below.

<sub>🤖 M1mk</sub>

File: src/ci.rs [hunk 7]
```diff
@@ -1779,6 +1836,153 @@ mod tests {
...
+        // Some("") is treated the same as None — no empty fenced block.
+        let checks = vec![CheckRun {
+            name: "CI / test".to_string(),
+            status: CheckStatus::Completed,
+            conclusion: Some(CheckConclusion::Failure),
+            duration: None,
+            output: Some("".to_string()),
+        }];
+        let prompt = build_ci_fix_prompt(&checks, 1);
+        assert!(
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/ci.rs:1875]: Copilot, fixed. Added `let dir = tempfile::tempdir().unwrap();` and changed the call to `build_ci_fix_prompt(&checks, 1, dir.path())`, matching the pattern used by all other tests in the module.

<sub>🤖 M1mk</sub>

File: src/ci.rs [hunk 8]
```diff
@@ -1779,6 +1836,153 @@ mod tests {
...
+    fn test_build_ci_fix_prompt_whitespace_output_shows_fallback() {
+        let checks = vec![CheckRun {
+            name: "CI / test".to_string(),
+            status: CheckStatus::Completed,
+            conclusion: Some(CheckConclusion::Failure),
+            duration: None,
+            output: Some("   \n  ".to_string()),
+        }];
+        let prompt = build_ci_fix_prompt(&checks, 1);
+        assert!(
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/ci.rs:1896]: Copilot, fixed. Same change as above: added `let dir = tempfile::tempdir().unwrap();` and passed `dir.path()` as the third argument.

<sub>🤖 M1mk</sub>
